groups/bsl/bslstl/bslstl_optional.h

```diff
@@ -448,6 +447,71 @@ struct Optional_Propagates_Allocator
           BloombergLP::bslstl::Optional_OptNoSuchType>::type =                \
           BloombergLP::bslstl::optNoSuchType
 
+#define BSLSTL_OPTIONAL_ENABLE_IF_CONSTRUCT_FROM_OPTIONAL_LVAL_DEF            \
+    , typename bsl::enable_if<                                                \
+          !bsl::is_same<ANY_TYPE, TYPE>::value &&                             \
+              !BloombergLP::bslstl::                                          \
+                  Optional_ConvertsFromOptional<TYPE, ANY_TYPE>::value &&     \
+              BloombergLP::bslstl::Optional_IsConstructible<TYPE,             \
+                                                            const ANY_TYPE&,  \
+                                                            true>::value,     \
+          BloombergLP::bslstl::Optional_OptNoSuchType>::type
+
```

Comment:
This new macros kinda pushed me over the edge.  Can we reduce how much macros actually contain and reduce the number of macros where possible?  I did a quick pass on this:

https://github.com/NinaRanns/bde/pull/4

I used the `shared_ptr` names `DECLARE_IF` and `DEFINE_IF` (instead of `MACRO` and `MACRO_DEF`).  

This also led me to a question if we're doing the right thing with `RVAL` stuff in C++03 - we're checking `Optional_IsConstructible(TYPE, ANY_TYPE`, but is this equivalent to `Optional_IsConstructible(TYPE, MovableRef<AnyType>` in C++03?  What do you think?

Also, I suggest removing putting the comma in the macro.  `shared_ptr` has a reason for it (macros expand to nothing in some cases), but I don't think this is a concern here.

Finally, the test driver didn't compile for me in VS2019 - I fixed it, but please double-check.

Reply:
Also, I'd try to avoid "implicit" arguments to the macro - e.g., `TYPE` and `ANY_TYPE` for what is currently `BSLSTL_OPTIONAL_ENABLE_IF_CONSTRUCT_FROM_OPTIONAL_LVAL`

Reply:
'MACRO' and 'MACRO_DEF' replaced with 'DECLARE_IF' and 'DEFINE_IF'

is_construcible trait doesn't exist in C++03, and Optional_IsConstructible in C++03 has the value of the bool given to it.

Reply:
removed commas from macros

Reply:
removed rval and lval macros

Reply:
all default arguments removed

Reply:
There's one thing that I hoped that we could do - we have `DECLARE_IF` and `DEFINE_IF` macros that have fairly complex conditions.  Can we reduce that duplication in the way that I suggested in #4 ? Also, please group `DECLARE_IF` with `DEFINE_IF` so the reader doesn't need to go up and down to confirm that the conditions are the same? I mean change from

```
DECLARE_IF_A
DECLARE_IF_B
DECLARE_IF_C
DEFINE_IF_A
DEFINE_IF_B
DEFINE_IF_C
```

to

```
DECLARE_IF_A
DEFINE_IF_A
DECLARE_IF_B
DEFINE_IF_B
DEFINE_IF_C
DECLARE_IF_C
```

Reply:
I have removed duplication in the LVAL/RVAL macros as suggested in your change. Am I missing something ?

Reply:
Sort of.  Here's what I mean.  We have the following two:

```
#define BSLSTL_OPTIONAL_DECLARE_IF_CONSTRUCTS_FROM(TYPE, ANYTYPE)             \
    typename bsl::enable_if<                                                  \
        !bsl::is_same<ANY_TYPE, TYPE>::value &&                               \
            !bsl::is_same<typename BloombergLP::bslstl::Optional_RemoveCVRef< \
                              ANY_TYPE>::type,                                \
                          bsl::optional<TYPE> >::value &&                     \
            !bsl::is_same<typename BloombergLP::bslstl::Optional_RemoveCVRef< \
                              ANY_TYPE>::type,                                \
                          bsl::nullopt_t>::value &&                           \
            !bsl::is_same<typename BloombergLP::bslstl::Optional_RemoveCVRef< \
                              ANY_TYPE>::type,                                \
                          bsl::in_place_t>::value &&                          \
            !bsl::is_same<typename BloombergLP::bslstl::Optional_RemoveCVRef< \
                              ANY_TYPE>::type,                                \
                          bsl::allocator_arg_t>::value &&                     \
            BloombergLP::bslstl::                                             \
                Optional_IsConstructible<TYPE, ANY_TYPE, true>::value,        \
        BloombergLP::bslstl::Optional_OptNoSuchType>::type =                  \
        BloombergLP::bslstl::optNoSuchType
```

```
#define BSLSTL_OPTIONAL_DEFINE_IF_CONSTRUCTS_FROM(TYPE, ANYTYPE)              \
    typename bsl::enable_if<                                                  \
        !bsl::is_same<ANY_TYPE, TYPE>::value &&                               \
            !bsl::is_same<typename BloombergLP::bslstl::Optional_RemoveCVRef< \
                              ANY_TYPE>::type,                                \
                          bsl::optional<TYPE> >::value &&                     \
            !bsl::is_same<typename BloombergLP::bslstl::Optional_RemoveCVRef< \
                              ANY_TYPE>::type,                                \
                          bsl::nullopt_t>::value &&                           \
            !bsl::is_same<typename BloombergLP::bslstl::Optional_RemoveCVRef< \
                              ANY_TYPE>::type,                                \
                          bsl::in_place_t>::value &&                          \
            !bsl::is_same<typename BloombergLP::bslstl::Optional_RemoveCVRef< \
                              ANY_TYPE>::type,                                \
                          bsl::allocator_arg_t>::value &&                     \
            BloombergLP::bslstl::                                             \
                Optional_IsConstructible<TYPE, ANY_TYPE, true>::value,        \
        BloombergLP::bslstl::Optional_OptNoSuchType>::type
```

My suggestion is to remove the duplication by having a `struct` with all the conditions spelled once inside and having macros using that struct as they do in #4 (see changes right at the beginning of the file).

Reply:
oh, I see. Sure. I'll do that and re-arrange the macros as suggested.

Reply:
changes applied. Note that the changes are slightly different than the suggested change as one can not disable functions based on the suggested `Optional_EnableIfConstructsFromOptional::type` (in cases where the condition is not satisfied `typename BloombergLP::bslstl::Optional_EnableIfConstructsFromOptional<TYPE, ANY_TYPE>::type` does not exist and the compilation fails). The tests in your branch only compile because they do not attempt to instantiate any of the constructors. 

This does mean that the new change is not as "slimline" as your suggestion. If it is deemed to be not sufficiently slimline, I could perhaps look into making a more clever ENABLE_IF macro that would take out the `typename bsl::enable_if<...>::type` boilerplate. WDYT ?